pkg/extendapi/devicetemplate.go

```diff
@@ -0,0 +1,216 @@
+)
+
+type templateData struct {
+	Template struct {
+		Metadata struct {
+			Labels map[string]string `yaml:"labels"`
+		} `yaml:"metadata"`
+	} `yaml:"template"`
+}
+
+type CreateDeviceTemplateHealthHandler struct {
+	clientset *kubernetes.Clientset
+	dyclient  dynamic.Interface
+}
+
+func NewCreateDeviceTemplateHandler(client *kubernetes.Clientset, dyclient dynamic.Interface) *CreateDeviceTemplateHealthHandler {
+	return &CreateDeviceTemplateHealthHandler{
+		clientset: client,
+		dyclient:  dyclient,
+	}
+}
+
+func (h *CreateDeviceTemplateHealthHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler request body error:", err.Error)
+		w.Write([]byte(fmt.Sprintf(`{"error":%s}`, err.Error())))
+		return
+	}
+
+	if !gjson.ValidBytes(body) {
+		logrus.Error("CreateDeviceTemplateHealthHandler request body valid json error")
+		w.Write([]byte(`{"error":"request body error"}`))
+		return
+	}
+
+	if !gjson.GetBytes(body, "deviceType").Exists() {
+		logrus.Error("CreateDeviceTemplateHealthHandler request body have no field deviceType")
+		w.Write([]byte(`{"error":"request body have no field deviceType"}`))
+		return
+	}
+
+	if !gjson.GetBytes(body, "user").Exists() {
+		logrus.Error("CreateDeviceTemplateHealthHandler request body have no field user")
+		w.Write([]byte(`{"error":"request body have no field user"}`))
+		return
+	}
+
+	if !gjson.GetBytes(body, "template").Exists() {
+		logrus.Error("CreateDeviceTemplateHealthHandler request body have no field template")
+		w.Write([]byte(`{"error":"request body have no field template"}`))
+		return
+	}
+
+	deviceType := gjson.GetBytes(body, "deviceType").String()
+	user := gjson.GetBytes(body, "user").String()
+	templateBase64 := gjson.GetBytes(body, "template").String()
+
+	templateBytes, err := base64.StdEncoding.DecodeString(templateBase64)
+	if err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler request tmplate base64 decode error:", err.Error)
+		w.Write([]byte(fmt.Sprintf(`{"error":%s}`, err.Error())))
+		return
+	}
+
+	if err := h.validTemplate(req.Context(), deviceType, templateBytes); err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler validTemplate error:", err.Error())
+		w.Write([]byte(fmt.Sprintf(`{"error":%s}`, err.Error())))
+		return
+	}
+
+	if err := h.createSecret(req.Context(), deviceType, user, templateBytes); err != nil {
+		logrus.Error("CreateDeviceTemplateHealthHandler createSecret error:", err.Error())
+		w.Write([]byte(fmt.Sprintf(`{"error":%s}`, err.Error())))
+		return
+	}
+
+	w.Write([]byte(fmt.Sprintf(`{"error":success}`)))
+}
+
+func (h *CreateDeviceTemplateHealthHandler) createSecret(ctx context.Context, deviceType, user string, templateBytes []byte) error {
+
+	tempStr := util.GenerateRandomTempKey(7)
+
+	deviceType = strings.ToLower(deviceType)
+
+	name := fmt.Sprintf("%s-template-%s", deviceType, tempStr)
```

Comment:
you should either add a display name of the deivceProperty struct or use the user's input name as the secret name.